src/core/net/srp_client.cpp

```diff
@@ -0,0 +1,1559 @@
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "srp_client.hpp"
+
+#include "common/code_utils.hpp"
+#include "common/debug.hpp"
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+#include "common/logging.hpp"
+#include "common/settings.hpp"
+#include "common/string.hpp"
+
+#if OPENTHREAD_CONFIG_SRP_CLIENT_ENABLE
+
+/**
+ * @file
+ *   This file implements the SRP client.
+ */
+
+namespace ot {
+namespace Srp {
+
+//---------------------------------------------------------------------
+// Client::TxtEntry
+
+otError Client::TxtEntry::AppendTo(Message &aMessage) const
+{
+    otError error = OT_ERROR_NONE;
+    uint8_t length;
+
+    if (mKey == nullptr)
+    {
+        VerifyOrExit(mValue != nullptr);
+        error = aMessage.AppendBytes(mValue, mValueLength);
+        ExitNow();
+    }
+
+    length = static_cast<uint8_t>(StringLength(mKey, kMaxKeyLength + 1));
+
+    VerifyOrExit(length <= kMaxKeyLength, error = OT_ERROR_INVALID_ARGS);
+
+    if (mValue == nullptr)
+    {
+        // Treat as a boolean attribute and encoded as "key" (with no `=`).
+        SuccessOrExit(error = aMessage.Append(length));
+        error = aMessage.AppendBytes(mKey, length);
+        ExitNow();
+    }
+
+    // Treat as key/value and encode as "key=value", value may be empty.
+
+    VerifyOrExit(mValueLength + length + sizeof(char) <= kMaxKeyValueEncodedSize, error = OT_ERROR_INVALID_ARGS);
+
+    length += static_cast<uint8_t>(mValueLength + sizeof(char));
+
+    SuccessOrExit(error = aMessage.Append(length));
+    SuccessOrExit(error = aMessage.AppendBytes(mKey, length));
+    SuccessOrExit(error = aMessage.Append<char>(kKeyValueSeparator));
+    error = aMessage.AppendBytes(mValue, mValueLength);
+
+exit:
+    return error;
+}
+
+otError Client::TxtEntry::AppendEntries(const TxtEntry *aEntries, uint8_t aNumEntries, Message &aMessage)
+{
+    otError  error       = OT_ERROR_NONE;
+    uint16_t startOffset = aMessage.GetLength();
+
+    for (uint8_t index = 0; index < aNumEntries; index++)
+    {
+        SuccessOrExit(error = aEntries[index].AppendTo(aMessage));
+    }
+
+    if (aMessage.GetLength() == startOffset)
+    {
+        uint8_t len = 0;
+
+        error = aMessage.Append(len);
+    }
+
+exit:
+    return error;
+}
+
+//---------------------------------------------------------------------
+// Client::HostInfo
+
+void Client::HostInfo::Init(void)
+{
+    Clearable<HostInfo>::Clear();
+
+    // State is directly set on `mState` instead of using `SetState()`
+    // to avoid logging.
+    mState = OT_SRP_CLIENT_ITEM_STATE_REMOVED;
+}
+
+void Client::HostInfo::Clear(void)
+{
+    Clearable<HostInfo>::Clear();
+    SetState(kRemoved);
+}
```

Comment:
I considered this but I prefer to state order we have (specially since we use tables/arrays to transition between states in multiple places). 

Also in case of `Clear()` we do want to call `SetState(kRemoved)` so that the state change is logged.